genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+ * missing entries. It uses OS level locks to accomplish the following
+ * <p>
+ * Allows multiple agents reading/writing/deleting in the same cache folder
+ * without corrupting the cache
+ * Downloads new versions of resources once they are available remotely
+ * Deletes the old versions of resources once a new version is successfully downloaded
+ * Recovers from partial downloads of resources in case an agent gets killed in the middle of a download
+ * or a download fails for any other reason
+ * <p>
+ * Does NOT handle concurrency within the same agent (not an issue at the moment)
+ * <p>
+ * Cache structure on local disk
+ * Each resource has a hash to represent it. The version of the resource is the remote last modified
+ * timestamp. For each resource and version, a lock file is created. Each process takes a lock on
+ * this file and then downloads the resource to a data.tmp file. Once the download is complete
+ * data.tmp is atomically moved to data.
+ * For e.g for a hash value of 6d331abc92bc8244bc5d41e2107f303a and last modified = 1525456404
+ * cache entries would like the following
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/data
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/lock
+ * <p>
+ * Deletion of older versions
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Service
+@Lazy
+@Slf4j
+@Component
+class FileSystemFetchingCacheService implements FetchingCacheService {
+
+    public static final String LOCK_FILE_NAME = "lock";
+    public static final String DATA_FILE_NAME = "data";
+    public static final String DATA_DOWNLOAD_FILE_NAME = "data.tmp";
+    protected final ResourceLoader resourceLoader;
+    protected final File cacheDirectory;
+    protected final LockingFactory lockingFactory;
+    private final ExecutorService oldResourcesCleanUpService = Executors.newFixedThreadPool(1);
+
+
+    FileSystemFetchingCacheService(
+        final ResourceLoader resourceLoader,
+        final ArgumentDelegates.CacheArguments cacheArguments,
+        final LockingFactory lockingFactory
+    ) {
+        this.resourceLoader = resourceLoader;
+
+        this.cacheDirectory = cacheArguments.getCacheDirectory();
+
+        this.lockingFactory = lockingFactory;
+
+        createDirectoryStructureIfNotExists(cacheDirectory);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final URI sourceFileUri, final File destinationFile) throws DownloadException, IOException {
+        lookupOrDownload(sourceFileUri, destinationFile);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final Set<Pair<URI, File>> sourceDestinationPairs) throws DownloadException, IOException {
+        for (final Pair<URI, File> sourceDestinationPair : sourceDestinationPairs) {
+            get(sourceDestinationPair.getKey(), sourceDestinationPair.getValue());
+        }
+    }
+
+    protected void lookupOrDownload(
```

Comment:
This is a pretty long method. You might want to break it up into smaller chunks via helper methods to make it easier to understand everything going on

Reply:
Ok